src/storage/database.rs

```diff
@@ -0,0 +1,1483 @@
+    pub successful: bool,
+    /// How long the decoding took in milliseconds
+    pub execution_time_ms: i64,
+    /// When the decoding was run
+    pub timestamp: String,
+}
+
+impl PartialEq for CacheRow {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id
+            && self.encoded_text == other.encoded_text
+            && self.decoded_text == other.decoded_text
+            && self.path == other.path
+            && self.successful == other.successful
+            && self.execution_time_ms == other.execution_time_ms
+            && self.timestamp == other.timestamp
+    }
+}
+
+#[derive(Debug)]
+/// Represents an entry into the cache table
+pub struct CacheEntry {
+    /// Text before it is decoded
+    pub encoded_text: String,
+    /// Text after it is decoded
+    pub decoded_text: String,
+    /// Ordered list of decoding attempts
+    pub path: Vec<CrackResult>,
+    /// How long the decoding took in milliseconds
+    pub execution_time_ms: i64,
+}
+
+/// Helper function get a DateTime formatted timestamp
+fn get_timestamp() -> String {
+    let timestamp: DateTime<chrono::Local> = std::time::SystemTime::now().into();
+    timestamp.format("%Y-%m-%d %T").to_string()
+}
+
+/// Returns the path to the database file
+fn get_database_path() -> std::path::PathBuf {
+    let mut path = dirs::home_dir().expect("Could not find home directory");
+    path.push(".ciphey");
+    path.push("database.sqlite");
+    path
+}
+
+/// Opens and returns a Connection to the SQLite database
+///
+/// If a path is specified in DB_PATH, returns a Connection to that path
+/// Otherwise, opens a Connection to an in-memory database
+fn get_db_connection() -> Result<rusqlite::Connection, rusqlite::Error> {
+    match DB_PATH.get() {
+        Some(Some(path)) => rusqlite::Connection::open(path),
+        _ => rusqlite::Connection::open_in_memory(),
+    }
+}
+
+/// Public wrapper for setting up database
+///
+/// # Errors
+///
+/// On error setting up the database, returns a rusqlite::Error
+/// If there's an error while setting the database path, prints warning
+/// to console and continues with the default DB_PATH
+pub fn setup_database() -> Result<(), rusqlite::Error> {
+    match DB_PATH.get() {
+        Some(_path) => (),
+        None => {
+            let db_result: Result<(), Option<std::path::PathBuf>> =
+                DB_PATH.set(Some(get_database_path()));
+            match db_result {
+                Ok(_) => (),
+                Err(_e) => {
+                    crate::cli_pretty_printing::warning("Error setting database path");
+                }
+            }
+        }
+    };
+    init_database()?;
+    Ok(())
+}
+
+/// Initializes database with default schema
+fn init_database() -> Result<rusqlite::Connection, rusqlite::Error> {
+    let conn = get_db_connection()?;
+    // Initializing cache table
+    conn.execute(
+        "CREATE TABLE IF NOT EXISTS cache (
+            id INTEGER PRIMARY KEY AUTOINCREMENT,
```

Comment:
when we merge DBs the auto increment will cause issues. In production DBs I think we do UUID for ID to prevent this but we can handle this in our merging scripts too